text/0000-secret-types.md

```diff
@@ -0,0 +1,206 @@
+
+These integers are intended to form a basis for the creation of other secret types, which can be built on top of them, such as secret strings for storing passwords in them.
+
+Comparison of secret integer types is allowed, but the algorithm is independent of the secrets. Comparison of secret types must return a secret_bool, which cannot be branched on.
+
+Example: comparison
+```
+let x : secret_i8 = 6;
+Let y : secret_i8 = 10;
+if (x < y) { // compiler error: cannot branch on secret bool
+  ...
+}
+```
+Example: declassification
+```
+let x : secret_i8 = 6;
+let y : secret_i8 = 10;
+println!((x ^ y).declassify())
+```
+
+Since indexing vectors and arrays is only possible using `usize,` you will be prevented from indexing into a vector or array with a secret integer. Error messages will derive naturally from the type system.
+
+
+Secret types will likely be a more advanced topic when teaching Rust.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+For each fixed-size integer type:
+Implement the following methods:
+- From_be
+- From_le
+- From_be_bytes
+- From_le_bytes
+- From_ne_bytes
+- Is_positive
+- Is_negative
+- Leading_zeros
+- Min_value
+- Max_value
+- Overflowing_add
+- Overflowing_sub
+- Overflowing_mul
+- Overflowing_neg
+- Overflowing_shl
+- Overflowiing_shr
+- Overflowing_pow
+- Pow
+- Reverse_bits
+- Rotate_left
+- Rotate_right
+- Saturating_add
+- Saturating_sub
+- Saturating_neg
+- Saturating_mul
+- Saturating_pow
+- Signum
+- Swap_bytes
+- To_be
+- To_le
+- To_be_bytes
+- To_le_bytes
+- To_ne_bytes
+- Trailing_zeros
+- Wrapping_add
+- Wrapping_sub
+- Wrapping_mul
+- Wrapping_neg
+- Wrapping_shl
+- Wrapping_shr
+- Wrapping_pow
+
+
+Implement the following traits
+Secret integers may only  be combined with other secret integers and the result will be a secret type
+- Add
+- AddAssign
+- BitAnd
+- BitAndAssign
+- BitOr
+- BitOrAssign
+- BitXor
+- BitXorAssign
+- Clone
+- Copy
+- Default
+- Drop
+- Mul
+- MulAssign
+- Neg
+- Not
+- Shl
+- ShlAssign
+- Shr
+- ShrAssign
+- Sub
+- SubAssign
+- Sub
+
+
+For secret_bool, implement the following methods:
+- n/a
+And the following traits
+- BitAnd
+- BitAndAssign
+- BitOr
+- BitOrAssign
+- BitXor
+- BitXorAssign
+- Clone
+- Copy
+- Default
+- Not
+
+We will also need to define a trait Classify<T> for T and method declassify. Classify will take a non secret integer or boolean and return a secret integer or boolean. `declassify` will consume a secret integer returning a non-secret integer.
+
+We will need to define new comparison traits that run in constant time and return a secret_bool:
+- SecretEq
+- SecretOrd
+For reference, see the subtle crate [2] .
+
+[2] https://docs.rs/subtle/2.2.2/subtle/
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Because secret integers prohibit the use of certain operations and compiler optimizations, there will be a performance impact on code that uses them. However, this is intentional.
```

Comment:
This is vague. Please be as specific as possible regarding what optimizations or categories of optimizations would be prohibited, at what granularity they would be prohibited, and what the impacts of those prohibitions would be.

Please also, without referring to LLVM, specify how these prohibitions are justified in terms of a specification. If *guarantees* (with stability promises and all that, as opposed to best effort constructs) are sought after, such a specification should be operational (in the sense of [operational semantics](https://internals.rust-lang.org/t/operational-semantics-and-high-level-vs-low-level/11639/5) and abstract machines). At the moment, I do not see how this can be specified in terms of an abstract machine / interpreter like Miri.

Reply:
I imagine one could re-use some [ideas from this paper](https://eprint.iacr.org/2019/926.pdf) and more generally from the IFC literature. The guarantee should probably be some form of *non-interference* -- program executions that only differ in secrets must not be observably different, or so. One hard problem is defining what exactly the possible observations are.

A type-based approach certainly sounds like a good start for me; this feels much better than [ad-hoc attributes](https://internals.rust-lang.org/t/annotations-for-zeroing-the-stack-of-sensitive-functions-which-deal-in-transient-secrets).

Reply:
Thanks for the paper reference; that's a good start. Rather than just enumerate possible observations, we can also enumerate observation modalities that are out of scope. My list of the latter includes 
1) instantaneous power fluctuations due to Hamming weight of secret data, whether on-chip or at interconnects between chips (e.g., CPU and RAM); and 
2) secret-influenced variation in RF radiation as secret data propagates through the circuitry.

Reply:
I think you made a very good point for why we need to precisely list what we *do* consider an observation -- everything not on that list (and that will be all sorts of weird stuff, like your two items) is not protected.

Reply:
I think you can generally put all microarchitectural covert channels out-of-scope, with the possible caveat that writing constant-time code is the best known defense against them.

That said I think compiling a list of more general architecture properties is a good idea. An example of one is: the underlying architecture is assumed to be able to perform integer multiplication in constant time (e.g. doesn't short-circuit on multiply-by-0-or-1)

Reply:
> That said I think compiling a list of more general architecture properties is a good idea. An example of one is: the underlying architecture is assumed to be able to perform integer multiplication in constant time (e.g. doesn't short-circuit on multiply-by-0-or-1)

I don't think such a strong assumption is required. Once the information that inputs to a multiplication are secret is forwarded to the compiler, it's the compiler's job to generate constant-time code. If there *is* a constant-time multiplier in the target micro-architecture, that's easy. If not, what the compiler will need are constant-time runtime libraries to use for arithmetic on secret integers. The critical job for the language to take care of is to forward the information about secrecy in the first place.

Reply:
>  If there is a constant-time multiplier in the target micro-architecture, that's easy. If not, what the compiler will need are constant-time runtime libraries to use for arithmetic on secret integers. 

Even across the same microarchitecture there are CPUs that do-or-do-not perform multiply in constant time, e.g. the PowerPC 7xx CPUs used by Apple perform constant time multiplication, but many other PPC32 CPUs do not.

Reply:
I agree that there is a difference between different microarchitectures (i.e., implementations of some architecture) with only some having variable-time arithmetic instructions. The situation is even worse if you take future implementations of some architecture into account, that suddenly introduce new variable-time behavior. Compilers will need a new hardware-software contract to know what subset of instructions is safe to use (see https://ts.data61.csiro.au/publications/csiro_full_text/Ge_YH_18.pdf). 

However, to me this all seems like lower-level issues: what can be done on language-level is to ensure that some operations are avoided on secret data that a compiler cannot possibly protect (branching, addressing) and for all other operations leave it to the compiler and hardware to guarantee constant-time behavior.

Reply:
Nice find! Here's a more recent one: https://arxiv.org/abs/1901.08338. 

One of the authors of that paper also contributed to this recent attack, which shows that Intel's attempt at ad-hoc patching is useless: https://cacheoutattack.com/

RIDL is another good attack, well written paper: https://mdsattacks.com/

Reply:
> I imagine one could re-use some [ideas from this paper](https://eprint.iacr.org/2019/926.pdf) and more generally from the IFC (information flow control) literature. The guarantee should probably be some form of _non-interference_ -- program executions that only differ in secrets must not be observably different, or so. One hard problem is defining what exactly the possible observations are; that can and should (IMO) be done in an operational way (a "trace of observations" that e.g. Miri could print, or so).

Co-author of said paper here. 

Cryptographic constant-time security is indeed usually stated as a non-interference property. This works by taking the operational semantics of the considered language, and instrumenting it by adding "leakages" or "observations" to step taken by instructions that can be compiled to some sort of jump or memory accesses. A program is then said to be secure if for two executions that differ only on secret inputs, then the leakages are equal.

As for impacted optimizations, surprisingly, the only thing impacted during our work on CompCert was the generation of builtin code that was not branchless, so it might be safe to consider those optimizations implemented in CompCert to not be impacted for llvm too: tailcall, inlining, CSE, constant propagation, deadcode elimination. That remains a lot less than the optimizations implemented in llvm though, I assume...